apps/src/applab/EventSandboxer.js

```diff
@@ -0,0 +1,162 @@
+/** @file Helper that generates sanitized, standardized event objects. */
+// Strict linting: Absorb into global config when possible
+/* jshint
+ unused: true,
+ eqeqeq: true,
+ maxlen: 120
+ */
+'use strict';
+
+var keyEvent = require('./keyEvent');
+
+/**
+ * Helper that generates sanitized, standardized event objects.
+ *
+ * Primary responsibilities include:
+ * - Allowing only a subset of properties through to sandboxed events
+ * - Replacing DOM references with string element IDs for security
+ * - Transforming screen coordinates to the app coordinate system
+ * - Polyfill movementX and movementY properties on mouse events
+ *
+ * @constructor
+ * @example
+ *   var sandboxer = new EventSandboxer();
+ *   function onClick(e) {
+ *     studentClickHandler(sandboxer.sandboxEvent(e));
+ *   }
+ */
+var EventSandboxer = module.exports = function () {
+  /** @private {number} */
+  this.xScale_ = 1;
+
+  /** @private {number} */
+  this.yScale_ = 1;
+
+  /** @private {number} */
+  this.xOffset_ = 0;
+
+  /** @private {number} */
+  this.yOffset_ = 0;
+
+  /**
+   * @private {Object} map from element id to the last mousemove event which
+   * occurred on that element. This is used to simulate movementX and movementY
+   * for browsers which do not support it natively.
+   */
+  this.lastMouseMoveEventMap_ = {};
+};
+
+/**
+ * Updates the sandboxer's internal transform according to the position and
+ * scale of the provided div (which should be the app/play space) so that it
+ * can accurately transform mouse coordinates into app space.
+ * @param {HTMLDivElement} div
+ */
+EventSandboxer.prototype.setTransformFromDiv = function (div) {
+  // Capture div scale
+  this.xScale_ = div.getBoundingClientRect().width / div.offsetWidth;
+  this.yScale_ = div.getBoundingClientRect().height / div.offsetHeight;
+
+  // Capture div offset, and all of its parent offsets, together
+  var xOffset = 0;
+  var yOffset = 0;
+  while (div) {
+    xOffset += div.offsetLeft;
+    yOffset += div.offsetTop;
+    div = div.offsetParent;
```

Comment:
Technically if we keep going upwards from divApplab, we eventually hit a non-div in body. This is mostly a nit.